sycl/include/sycl/builtins.hpp

```diff
@@ -602,207 +599,261 @@ __SYCL_MATH_FUNCTION_3_OVERLOAD(mad) __SYCL_MATH_FUNCTION_3_OVERLOAD(mix)
+  }
+    __SYCL_DEF_BUILTIN_GENFLOAT
+#undef __SYCL_BUILTIN_DEF
 
-// svgenfloat fmax (svgenfloat x, sgenfloat y)
-template <typename T>
-std::enable_if_t<detail::is_vgenfloat<T>::value, T>
-fmax(T x, typename T::element_type y) __NOEXC {
-  return __sycl_std::__invoke_fmax<T>(x, T(y));
-}
+// genfloat fmax (genfloat x, genfloat y)
+#define __SYCL_BUILTIN_DEF(TYPE)                                               \
+  inline TYPE fmax(TYPE x, TYPE y) __NOEXC {                                   \
+    return __sycl_std::__invoke_fmax<TYPE>(x, y);                              \
+  }
+    __SYCL_DEF_BUILTIN_GENFLOAT
+#undef __SYCL_BUILTIN_DEF
 
-// svgenfloat fmin (svgenfloat x, svgenfloat y)
-template <typename T>
-std::enable_if_t<detail::is_svgenfloat<T>::value, T> fmin(T x, T y) __NOEXC {
-  return __sycl_std::__invoke_fmin<T>(x, y);
-}
+// genfloat fmax (genfloat x, sgenfloat y)
+#define __SYCL_BUILTIN_DEF(TYPE)                                               \
+  inline TYPE fmax(TYPE x, TYPE::element_type y) __NOEXC {                     \
+    return __sycl_std::__invoke_fmax<TYPE>(x, TYPE(y));                        \
+  }
+    __SYCL_DEF_BUILTIN_VGENFLOAT
+#undef __SYCL_BUILTIN_DEF
 
-// svgenfloat fmin (svgenfloat x, sgenfloat y)
-template <typename T>
-std::enable_if_t<detail::is_vgenfloat<T>::value, T>
-fmin(T x, typename T::element_type y) __NOEXC {
-  return __sycl_std::__invoke_fmin<T>(x, T(y));
-}
+// genfloat fmin (genfloat x, genfloat y)
+#define __SYCL_BUILTIN_DEF(TYPE)                                               \
+  inline TYPE fmin(TYPE x, TYPE y) __NOEXC {                                   \
+    return __sycl_std::__invoke_fmin<TYPE>(x, y);                              \
+  }
+    __SYCL_DEF_BUILTIN_GENFLOAT
+#undef __SYCL_BUILTIN_DEF
 
-// svgenfloat fmod (svgenfloat x, svgenfloat y)
-template <typename T>
-std::enable_if_t<detail::is_svgenfloat<T>::value, T> fmod(T x, T y) __NOEXC {
-  return __sycl_std::__invoke_fmod<T>(x, y);
-}
+// genfloat fmin (genfloat x, sgenfloat y)
+#define __SYCL_BUILTIN_DEF(TYPE)                                               \
+  inline TYPE fmin(TYPE x, TYPE::element_type y) __NOEXC {                     \
+    return __sycl_std::__invoke_fmin<TYPE>(x, TYPE(y));                        \
+  }
+    __SYCL_DEF_BUILTIN_VGENFLOAT
+#undef __SYCL_BUILTIN_DEF
 
-// svgenfloat fract (svgenfloat x, genfloatptr iptr)
-template <typename T, typename T2>
-std::enable_if_t<
-    detail::is_svgenfloat<T>::value && detail::is_genfloatptr<T2>::value, T>
-fract(T x, T2 iptr) __NOEXC {
+// genfloat fmod (genfloat x, genfloat y)
+#define __SYCL_BUILTIN_DEF(TYPE)                                               \
+  inline TYPE fmod(TYPE x, TYPE y) __NOEXC {                                   \
+    return __sycl_std::__invoke_fmod<TYPE>(x, y);                              \
+  }
+    __SYCL_DEF_BUILTIN_GENFLOAT
+#undef __SYCL_BUILTIN_DEF
+
+    // svgenfloat fract (svgenfloat x, genfloatptr iptr)
```

Comment:
I just noticed there are no some built-ins that accept arguments with gen*ptr:

> genfloat fract(genfloat x, genfloatptr iptr)
> genfloat frexp(genfloat x, genintptr exp)
> genfloat lgamma_r(genfloat x, genintptr signp)
> genfloat modf(genfloat x, genfloatptr iptr)
> genfloat remquo(genfloat x, genfloat y, genintptr quo)
> genfloat sincos(genfloat x, genfloatptr cosval)

I think it will be better to create a separate patch for them as this patch is already too big for review.